UofT-Frosh-Orientation/orientation-website
# 620 PDF data receipt

## **Contribution:**
Created a PDF using React-PDF with a unique QR Code based on MongoDB user id, allows to download from the profile page. 

Additionally, the PDF receipt is sent to the user upon registration by sending the Blob to the backend and sending an email containing the PDF.


### **Issue:**

Closes #620 

### **Accomplishments:**

- Loaded QR Code in React-PDF (oh my god)
- Created PDF Blob Provider in the Profile Page to Download PDF Receipt based on user information
- Added PDF Blob provider to the registration page
- Added Blob and registration data to a `FromData` object and send to the backend
- Handled file on backend using `multer`
- Created `sendRawEmailMulterFiles` in `EmailServices.js`

### **Visuals:** 
![image](https://github.com/UofT-Frosh-Orientation/orientation-website/assets/48883986/48fc7388-5001-4e14-80c5-8955c66cc8a9)


## **Details:**
The PDF is basically a dump of all registration information barring the "secret" information contained in the database.

The backend `multer` package handles FromData objects that have a file/files attached and turn them into a `Buffer` object. However when passing this `Buffer` object through functions it turns into a regular object with a `type` and `data` field. This needs to be turned back into a `Buffer` and turned into a base64 encoded string.

### **Expected Behavior:**

- There should be a button the profile page that can be used to generate a unique PDF with a unique QR Code that has the mongoDB id alongside the rest of the user information from registration
- Once registration is complete and the user presses the pay button an email will be send to them which contains the PDF.

### **Testing Steps:**

Steps required to get expected behavior.

- Start the docker containers as usual.
- Create a user / login to an existing user.
- Navigate to the Resources Section and find the last button.
- Click and it should open a new tab with the PDF opened as a blob, download if needed.
- Click registration and fill out form.
- Press pay at the end.
- Once you're at the payments page you should receive an email with the PDF.

### **Complications:**

- The QR Code is hellish, I can't even begin to put into words just how annoying that was to setup in the end
- The PDF looks horrible, styling will have to be implemented later
- Errors in the `MakeReceipt()` function can stop the creation of the PDF but can be hard to detect.
- Managing the `Buffer` object in the backend is also hellish.

### **Resources:**
- [react-qrbtf package Docs (they are not in English)](https://github.com/CPunisher/react-qrbtf) 
- [React-PDF Docs](https://react-pdf.org/)
- [mimemessage NPM page](https://www.npmjs.com/package/mimemessage)
- [AWS Email API Raw Email Docs](https://docs.aws.amazon.com/ses/latest/APIReference-V2/API_SendEmail.html)
